diff -ur unzip-5.51/unix/unix.c unzip-5.51-lhh/unix/unix.c
--- unzip-5.51/unix/unix.c	2004-04-11 20:29:58.000000000 -0400
+++ unzip-5.51-lhh/unix/unix.c	2004-11-22 14:34:21.468981000 -0500
@@ -1052,7 +1052,9 @@
             fclose(G.outfile);
             return;
         }
-        fclose(G.outfile);                  /* close "link" file for good... */
+
+	if (G.outfile)
+            fclose(G.outfile);         /* close "link" file for good... */
         slnk_entry->target[ucsize] = '\0';
         if (QCOND2)
             Info(slide, 0, ((char *)slide, "-> %s ",
